### PR TITLE
CRI: Support string user name.

### PR DESCRIPTION
**Pull request by Random-Liu**
https://github.com/kubernetes/kubernetes/pull/33239 and https://github.com/kubernetes/kubernetes/pull/34811 combined together broke the cri e2e test. https://k8s-testgrid.appspot.com/google-gce#gci-gce-cri

The reason is that:
1) In dockershim and dockertools, we assume that `Image.Config.User` should be an integer. However, sometimes when user build the image with `USER nobody:nobody` or `USER root:root`, the field will become `nobody:nobody` and `root:root`. This makes dockershim to always return error.
2) The new kube-dns-autoscaler image is using `USER nobody:nobody`. (See https://github.com/kubernetes-incubator/cluster-proportional-autoscaler/blob/master/Dockerfile.in#L21)

This doesn't break the normal e2e test, because in dockertools [we only inspect image uid if `RunAsNonRoot` is set](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/dockertools/docker_manager.go#L2333-L2338), which is just a coincidence. However, in kuberuntime, [we always inspect image uid first](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kuberuntime/kuberuntime_container.go#L141).

This PR adds literal `root` and `nobody` support. One problem is that `nobody` is not quite the same in different OS distros. Usually it should be `65534`, but some os distro doesn't follow that. For example, Fedora is using `99`. (See https://lists.fedoraproject.org/archives/list/devel@lists.fedoraproject.org/thread/Q5GCKZ7Q7PAUQW66EV7IBJGSRJWYXBBH/?sort=date)

Possible solution:
* Option 1: ~~Just use `65534`. This is fine because currently we only need to know whether the user is root or not.~~ Actually, we need to pass the user id to runtime when creating a container.
* Option 2: Return the uid as string in CRI, and let kuberuntime handle the string directly.

This PR is using option 1.

@yujuhong @feiskyer 
/cc @kubernetes/sig-node
/cc @MrHohn 

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/36423)

### CODE REVIEW

**Comment by Random-Liu** [comment]
@saad-ali This is an important bug fix for cri integration. :)

File: pkg/kubelet/dockertools/docker_manager.go [hunk 1]
```diff
@@ -2495,8 +2495,23 @@ func GetUidFromUser(id string) string {
...
+	case "root":
+		return "0"
+	case "nobody":
+		// On many os distros, nobody is default to 65534. However,
+		// **NOTICE** that this is not always true, for example on
+		// Fedora nobody is default to 99.
+		// This is fine for now, because we only need to know whether
+		// the user is root or not. However, in the future, this may
+		// become a problem if we rely on accurate uid.
+		return "65534"
```
Comment (feiskyer) [inline pkg/kubelet/dockertools/docker_manager.go]: Is possible to get nobody's uid from system,e.g. by `Lookup` from `os/user`?

**Comment by Random-Liu** [comment]
@k8s-bot cri test this.

File: pkg/kubelet/dockertools/docker_manager.go (hunk 1 above)
Comment (Random-Liu) [inline pkg/kubelet/dockertools/docker_manager.go]: Possible, will have a try tomorrow.

Comment (euank) [inline pkg/kubelet/dockertools/docker_manager.go]: That's not right though. The UID we need is the mapping from the container's rootfs, not from the host distro. e.g. nobody can and will have different uids on the host OS and the container image, and in this case the string "nobody" is meant to be resolved by the runtime using the image's information.

**Comment by euank** [comment]
I don't think this is the right fix.

Someone can set an arbitrary user string in a Dockerfile. E.g., the "jenkins" image on dockerhub runs as the "jenkins" user.

If the old runtime can run these images and CRI can't, that's a bigger bug, and adding a few new usernames won't help.

Am I misunderstanding something?

**Comment by yujuhong** [comment]
According to https://github.com/kubernetes/kubernetes/pull/34811#discussion_r86698869, we don't support the `run_as_non_root` check for non-numeric users. In that case, dockershim should return an empty user id for the image. Kubelet won't set the user id in the ContainerConfig so that docker will respect what's specified in the image. This follows most closely to the existing behavior.

An alternative is to return "user" as a string in the image spec. Kubelet can check if it's a numeric id. If yes, it'll apply the `run_as_non_root` check. 
Whether it is a numeric id or not, kubelet will pass this value to the runtime via CRI (`LinuxContainerSecurityContext`) and let the runtime use it. 

The latter approach allows non-numeric user string to be passed via CRI, while the first one rely on the implicit behavior of using the image default.
Thoughts?

**Comment by Random-Liu** [comment]
> I don't think this is the right fix. Someone can set an arbitrary user string in a Dockerfile. E.g., the "jenkins" image on dockerhub runs as the "jenkins" user.

Yeah, you are right. I didn't check Dockerfile carefully yestoday. :)

**Comment by Random-Liu** [comment]
Offline discussed with @yujuhong.

For option 1.
* We can't perform any check on empty UID.
* We will rely more on runtime's default behaviour - If no user is specified in container config, it should create container with default user specified in the image.

For option 2.
* We still can't perform check on string UID.
* We rely less on the runtime's default behaviour, kubelet will always explicitly set user id in the container config.

It seems that option 2 is a better solution here.

**Comment by saad-ali** [comment]
> @saad-ali This is an important bug fix for cri integration. :)

Ack. Ok for post-code freeze merge.

**Comment by Random-Liu** [comment]
@k8s-bot cri test this.

**Comment by Random-Liu** [comment]
@yujuhong @feiskyer The PR is ready for review.

This PR:
1) Changes the CRI to let runtime always return and accept `User` as string.
2) Change runtime to skip non-root verification if `User` is not numeric, but not throw out error.
3) Notice that if `User` returned by `ImageStatus` is empty string, kuberuntime will treat it as root.

/cc @kubernetes/sig-node

**Comment by yujuhong** [comment]
@k8s-bot cri test this

**Comment by yujuhong** [comment]
LGTM. Marking p1 to make the builds green.

**Comment by feiskyer** [comment]
👍

**Comment by jonboulle** [comment]
@Random-Liu can we add some more specific guidance about what exactly a container runtime is supposed to do with this field? I understand we're leaning on backwards-compatible behaviour, but right now it's rather underspecified/open to interpretation in the CRI itself (e.g. do I create a new user, use the container image's /etc/passwd, do an nsswitch lookup, copy from the host, ...). As an example, perhaps something like [this](https://github.com/appc/spec/blame/master/spec/aci.md#L231)?

**Comment by Random-Liu** [comment]
@jonboulle Good suggestion! Will take a look at the description in aci.

**Comment by yujuhong** [comment]
@jonboulle thanks for the suggestion! As you said, CRI as an API needs more clarification. I think we should work on this during the code freeze (which is now until the release time). Please do not hesitate to send PRs!

**Pull request merged**